### PR TITLE
tests: Introduce new Parallel Benchmark

### PR DESCRIPTION
**Pull request by def-**
Example run: https://buildkite.com/materialize/nightly/builds/9354#0191a800-eb79-45b2-b657-7a8f9990f3e5
### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered. ([trigger-ci for additional test/nightly runs](https://trigger-ci.dev.materialize.com/))
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] If this PR includes major [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note), I have pinged the relevant PM to schedule a changelog post.

### CODE REVIEW

**Comment by def-** [comment]
@sjwiesman @petrosagg @antiguru Do you think this is suitable to help with our efforts to improve and ensure performance? I want this to be usable for:
- local tests
- in Nightly CI to make sure we don't regress
- periodically against real Mz cloud instances to verify that our CI numbers are plausible
- as well as for some kind of customer-facing report of expected performance (table with scenarios and cc sizes, I'm still missing the scenarios we'd want for that).

It's still in a pretty early state, but I wanted to get some feedback before investing more time. An example run in CI can be seen in https://buildkite.com/materialize/nightly/builds/9352#0191a772-5c87-425c-b62a-1b54261a09c1 (expand groups), can be run locally with `bin/mzcompose --find open-bench run default --scenario=PoolRead` and can be run against an Mz cloud instance by adding a `--mz-url='postgres://...'`

File: test/parallel-benchmark/mzcompose.py [hunk 1]
```diff
@@ -0,0 +1,1186 @@
...
+    """Run the action in each thread in one second, spread apart by the 1/per_second"""
+
+    def __init__(self, per_second: int):
+        self.per_second = per_second
+
+    def generate(self, duration: int) -> Iterator[None]:
+        sleep_time = 1 / self.per_second
+        for i in range(duration * self.per_second):
+            yield
+            time.sleep(sleep_time)
```
Comment (antiguru) [inline test/parallel-benchmark/mzcompose.py:224]: This doesn't produce a true periodic distribution as it can be affected by scheduling. Instead, it should determine at which point it needs to produce a record, and sleep the right amount of time. This ensures that even under load, we'll produce requests that are in the limit evenly-spaced.

File: test/parallel-benchmark/mzcompose.py [hunk 2]
```diff
@@ -0,0 +1,1186 @@
...
+    def __init__(self, mean: float, stddev: float):
+        self.mean = mean
+        self.stddev = stddev
+
+    def generate(self, duration: int) -> Iterator[None]:
+        end_time = datetime.datetime.now() + datetime.timedelta(seconds=duration)
+        while datetime.datetime.now() < end_time:
+            yield
+            sleep_time = max(0, random.gauss(self.mean, self.stddev))
+            time.sleep(sleep_time)
```
Comment (antiguru) [inline test/parallel-benchmark/mzcompose.py:239]: Same as above, we need to sleep such that we're not affected by scheduling.

File: test/parallel-benchmark/mzcompose.py [hunk 3]
```diff
@@ -0,0 +1,1186 @@
...
+
+    def run(
+        self,
+        duration: int,
+        jobs: queue.Queue,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        for _ in self.dist.generate(duration):
+            jobs.put(lambda: self.action.run(conns, measurements))
```
Comment (antiguru) [inline test/parallel-benchmark/mzcompose.py:266]: This might not measure the desired latency. In open-loop scenarios, what we want is the time from request generation to fulfillment, but here we seem to measure the time from local dequeue until fulfillment, which can be considerably lower if the requests spends time in the local queue.

File: test/parallel-benchmark/mzcompose.py [hunk 4]
```diff
@@ -0,0 +1,1186 @@
...
+        return f"{self.timestamp} {self.value}"
+
+
+class Action:
+    def run(
+        self,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ):
+        start_time = datetime.datetime.now()
```
Comment (antiguru) [inline test/parallel-benchmark/mzcompose.py:126]: See comment further down, but we need to pass the start time into the `run` method to get realistic end-to-end measurements, instead of determining the start time here.

**Review by antiguru — CHANGES_REQUESTED** [review]
Looks like a start, but in the current form it won't measure the right end-to-end latency! But, I don't think it's far away. I left additional comments around the precision of the various tickers.

An additional though: The generators only yield one request per loop, which will limit the maximal throughput to how fast we can loop. Ideally, it'll generate a batch of requests to avoid the overhead of sleeping and looping.

**Comment by def-** [comment]
> An additional though: The generators only yield one request per loop, which will limit the maximal throughput to how fast we can loop. Ideally, it'll generate a batch of requests to avoid the overhead of sleeping and looping.

I'm expecting that for most interesting scenarios Materialize will be limiting the throughput.

Implemented your other suggestions

File: test/parallel-benchmark/mzcompose.py [hunk 5]
```diff
@@ -263,7 +268,8 @@ def run(
         measurements: collections.defaultdict[str, list[Measurement]],
     ) -> None:
         for _ in self.dist.generate(duration):
-            jobs.put(lambda: self.action.run(conns, measurements))
+            start_time = time.time()
+            jobs.put(lambda: self.action.run(start_time, conns, measurements))
```
Comment (antiguru) [inline test/parallel-benchmark/mzcompose.py:272]: We should even let `generate` yield the start time instead of capturing it here.

**Review by antiguru — APPROVED** [review]
Thank you, looks like what I was imagining!

File: test/parallel-benchmark/mzcompose.py [hunk 6]
```diff
@@ -0,0 +1,1190 @@
...
+class Periodic(Distribution):
+    """Run the action in each thread in one second, spread apart by the 1/per_second"""
+
+    def __init__(self, per_second: int):
+        self.per_second = per_second
+
+    def generate(self, duration: int) -> Iterator[float]:
+        next_time = time.time()
+        for i in range(duration * self.per_second):
+            yield time.time()
```
Comment (antiguru) [inline test/parallel-benchmark/mzcompose.py:226]: I think this should yield `next_time` instead of the current time, then we're actually modelling the external system correctly. The key insight is that we're never in control about when requests arrive, so we need to use `next_time`, which is a representative of the request produced by the external system.

File: test/parallel-benchmark/mzcompose.py [hunk 7]
```diff
@@ -0,0 +1,1190 @@
...
+
+    def connect(self) -> pg8000.Connection:
+        return pg8000.connect(
+            host=self.host,
+            port=self.port,
+            user=self.user,
+            password=self.password,
+            database=self.database,
+            ssl_context=ssl.SSLContext() if self.ssl else None,
+        )
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:89]: Shall we move this into some utils file? I guess we have something similar in several places.

File: test/parallel-benchmark/mzcompose.py [hunk 8]
```diff
@@ -0,0 +1,1190 @@
...
+
+
+def parse_pg_conn_string(conn_string: str) -> PgConnInfo:
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:92]: Maybe also something for utils

File: test/parallel-benchmark/mzcompose.py (hunk 7 above)
Comment (nrainer-materialize) [reply test/parallel-benchmark/mzcompose.py:89]: For example, `MzDbConfig` is also partially similar.

File: test/parallel-benchmark/mzcompose.py [hunk 9]
```diff
@@ -0,0 +1,1190 @@
...
+    )
+
+
+class Measurement:
+    value: float
+    timestamp: float
+
+    def __init__(self, value: float, timestamp: float | None = None):
+        self.value = value
+        self.timestamp = timestamp or time.time()
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:113]: Why is this necessary, why do we hand in a `timestamp` with `None`?

File: test/parallel-benchmark/mzcompose.py [hunk 10]
```diff
@@ -0,0 +1,1190 @@
...
+        ssl=query_params.get("sslmode", ["disable"])[-1] != "disable",
+    )
+
+
+class Measurement:
+    value: float
+    timestamp: float
+
+    def __init__(self, value: float, timestamp: float | None = None):
+        self.value = value
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:112]: Is this always a duration? Probably not but if so, can we rename the field? Otherwise, shall we store some information what it holds (e.g., the unit)?

File: test/parallel-benchmark/mzcompose.py [hunk 11]
```diff
@@ -0,0 +1,1190 @@
...
+        query: str,
+        conn_info: PgConnInfo,
+        strict_serializable: bool = True,
+    ):
+        self.query = query
+        self.conn = conn_info.connect()
+        self.conn.autocommit = True
+        self.cur = self.conn.cursor()
+        if not strict_serializable:
+            self.cur.execute("SET TRANSACTION_ISOLATION TO 'SERIALIZABLE'")
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:181]: Do we need to reset this if we reuse the connection?

File: test/parallel-benchmark/mzcompose.py [hunk 12]
```diff
@@ -0,0 +1,1190 @@
...
+        jobs: queue.Queue,
+        conns: queue.Queue,
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        c.testdrive(self.td, quiet=True)
+
+
+class LoadPhase(Phase):
+    duration: int
+    phase_actions: list[ClosedLoop | OpenLoop]
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:317]:
```suggestion
    phase_actions: list[PhaseAction]
```

File: test/parallel-benchmark/mzcompose.py [hunk 13]
```diff
@@ -0,0 +1,1190 @@
...
+        for thread in threads:
+            thread.start()
+        for i in range(self.duration):
+            assert (
+                jobs.qsize() < 100
+            ), "Too many jobs in queue, can't keep up: {jobs.qsize()}"
+            # if jobs.qsize() != 0:
+            #    measurements["Job queue"].append(Measurement(jobs.qsize()))
+            # if conns.qsize() != 0:
+            #    measurements["Connection queue"].append(Measurement(conns.qsize()))
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:347]: Commented out code...

File: test/parallel-benchmark/mzcompose.py [hunk 14]
```diff
@@ -0,0 +1,1190 @@
...
+            for phase_action in self.phase_actions
+        ]
+        for thread in threads:
+            thread.start()
+        for i in range(self.duration):
+            assert (
+                jobs.qsize() < 100
+            ), "Too many jobs in queue, can't keep up: {jobs.qsize()}"
+            # if jobs.qsize() != 0:
+            #    measurements["Job queue"].append(Measurement(jobs.qsize()))
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:345]: Optional: enum as key?

File: test/parallel-benchmark/mzcompose.py [hunk 15]
```diff
@@ -0,0 +1,1190 @@
...
+        measurements: collections.defaultdict[str, list[Measurement]],
+    ) -> None:
+        for phase in self.phases:
+            phase.run(c, self.jobs, self.conns, measurements)
+
+    def teardown(self) -> None:
+        while not self.conns.empty():
+            conn = self.conns.get()
+            conn.close()
+            self.conns.task_done()
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:414]: Is this correctly indented or does it need to be outside of the while block?

File: test/parallel-benchmark/mzcompose.py [hunk 16]
```diff
@@ -0,0 +1,1190 @@
...
+        for i in range(len(self.thread_pool)):
+            self.jobs.put(None)
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:416]: Why is this necessary? Add a comment?

File: test/parallel-benchmark/mzcompose.py [hunk 17]
```diff
@@ -0,0 +1,1190 @@
...
+                            ),
+                        )
+                        for i in range(10)
+                    ],
+                ),
+            ],
+            guarantees={
+                "SELECT * FROM kafka_mv (standalone)": {"qps": 15, "max": 10000},
+            },
+        )
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:482]: Nice approach!

File: test/parallel-benchmark/mzcompose.py [hunk 18]
```diff
@@ -0,0 +1,1190 @@
...
+                            ),
+                            dist=Periodic(per_second=125),
+                        ),
+                    ],
+                ),
+            ],
+            guarantees={
+                # TODO(def-): Lower max when RTR becomes more performant
+                "SET REAL_TIME_RECENCY TO TRUE; SELECT * FROM mv_sum (standalone)": {
+                    "qps": 50,
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:600]: qps is queries per second, right? Why is the value higher compared to `PgReadReplica`?
Resp. why "Lower max when RTR becomes more performant"? Am I misunderstanding that a higher value implies higher performance?

File: test/parallel-benchmark/mzcompose.py [hunk 19]
```diff
@@ -0,0 +1,1190 @@
...
+
+    def __init__(self, c: Composition, conn_infos: dict[str, PgConnInfo]):
+        self.init(
+            [
+                LoadPhase(
+                    duration=120,
+                    actions=[
+                        OpenLoop(
+                            action=ReuseConnQuery(
+                                "ALTER SYSTEM SET enable_table_keys = true",
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:736]: <s>What is the benefit of running this multiple times? Or was the intention to flip the flag?</s>
Understood after looking at the referenced issue.

File: test/parallel-benchmark/mzcompose.py [hunk 20]
```diff
@@ -0,0 +1,1190 @@
...
+            target = parse_pg_conn_string(args.mz_url)
+            c.up("testdrive", persistent=True)
+            conn_infos = {"materialized": target}
+            conn = target.connect()
+            with conn.cursor() as cur:
+                mz_version = cur.execute("SELECT mz_version()")
+                mz_version = cur.fetchall()[0][0]
+            conn.close()
+            mz_string = f"{mz_version} ({target.host})"
+    else:
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:1074]: What's the comparison target in this case? Or is no comparison performed?

File: test/parallel-benchmark/mzcompose.py [hunk 21]
```diff
@@ -0,0 +1,1190 @@
...
+        scenario_name = scenario_class.__name__
+        print(f"--- Running scenario {scenario_name}")
+        measurements = collections.defaultdict(list)
+        stats: dict[str, Statistics] = {}
+        scenario = scenario_class(c, conn_infos)
+        scenario.setup(c, conn_infos)
+        start_time = time.time()
+        Path(MZ_ROOT / "plots").mkdir(parents=True, exist_ok=True)
+        try:
+            # Don't let the garbage collector interfere with our measurements
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:1132]: I guess we need this for other benchmarks (feature benchmark, scalability framework) as well?

File: test/parallel-benchmark/mzcompose.py [hunk 22]
```diff
@@ -0,0 +1,1190 @@
...
+                                TestFailureDetails(
+                                    message=failure,
+                                    details=str(stats[key]),
+                                    test_class_name_override=scenario_name,
+                                )
+                            )
+                        else:
+                            print(
+                                f"Scenario {scenario_name} succeeded: {key}: {stat}: {value:.2f} {'>=' if less_than else '<='} {guarantee:.2f}"
+                            )
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:1163]: consider extracting a function

File: test/parallel-benchmark/mzcompose.py [hunk 23]
```diff
@@ -0,0 +1,1190 @@
...
+                print(f"Statistics for {key}:\n{stats[key]}")
+                if key in scenario.guarantees and args.guarantees:
+                    for stat, guarantee in scenario.guarantees[key].items():
+                        value = getattr(stats[key], stat)
+                        less_than = stat == "qps"
+                        if value < guarantee if less_than else value > guarantee:
+                            failure = f"Scenario {scenario_name} failed: {key}: {stat}: {value:.2f} {'<' if less_than else '>'} {guarantee:.2f}"
+                            print(failure)
+                            failures.append(
+                                TestFailureDetails(
```
Comment (nrainer-materialize) [inline test/parallel-benchmark/mzcompose.py:1154]: Nice!

File: test/parallel-benchmark/mzcompose.py (hunk 9 above)
Comment (def-) [reply test/parallel-benchmark/mzcompose.py:113]: Currently nowhere, was previously used in an older state

File: test/parallel-benchmark/mzcompose.py (hunk 10 above)
Comment (def-) [reply test/parallel-benchmark/mzcompose.py:112]: So far only durations.

File: test/parallel-benchmark/mzcompose.py (hunk 12 above)
Comment (def-) [reply test/parallel-benchmark/mzcompose.py:317]: mypy is a bit annoying here:
```
  /home/deen/git/materialize2/test/parallel-benchmark/mzcompose.py:415:29 - error: Argument of type "list[ClosedLoop | OpenLoop]" cannot be assigned to parameter "actions" of type "list[PhaseAction]" in function "__init__"
```
I don't want to list the type everywhere explicitly.

File: test/parallel-benchmark/mzcompose.py (hunk 15 above)
Comment (def-) [reply test/parallel-benchmark/mzcompose.py:414]: It is correct!

File: test/parallel-benchmark/mzcompose.py (hunk 12 above)
Comment (nrainer-materialize) [reply test/parallel-benchmark/mzcompose.py:317]: Ah, ok. You could avoid this issue by using a sequence instead of a list.

File: test/parallel-benchmark/mzcompose.py (hunk 20 above)
Comment (def-) [reply test/parallel-benchmark/mzcompose.py:1074]: No comparison performed, only checks the guarantees.

File: test/parallel-benchmark/mzcompose.py (hunk 18 above)
Comment (def-) [reply test/parallel-benchmark/mzcompose.py:600]: They run different things. The PgReadReplica scenario is closed loop with 10 threads in parallel, while the PgReadReplicaRTR is open loop with 125 queries per second.
